Common/src/main/java/com/blamejared/crafttweaker/api/tag/manager/ITagManager.java

```diff
@@ -1,187 +1,158 @@
 package com.blamejared.crafttweaker.api.tag.manager;
 
+import com.blamejared.crafttweaker.api.CraftTweakerAPI;
+import com.blamejared.crafttweaker.api.action.tag.ActionTagAdd;
+import com.blamejared.crafttweaker.api.action.tag.ActionTagCreate;
 import com.blamejared.crafttweaker.api.annotation.ZenRegister;
 import com.blamejared.crafttweaker.api.bracket.CommandStringDisplayable;
 import com.blamejared.crafttweaker.api.tag.MCTag;
 import com.blamejared.crafttweaker_annotations.annotations.Document;
+import net.minecraft.core.Holder;
+import net.minecraft.core.Registry;
+import net.minecraft.resources.ResourceKey;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.tags.Tag;
-import net.minecraft.tags.TagCollection;
+import net.minecraft.tags.TagManager;
 import org.openzen.zencode.java.ZenCodeType;
 
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Iterator;
 import java.util.List;
+import java.util.Map;
+import java.util.Objects;
 import java.util.stream.Collectors;
 
-/**
- * TagManagers are used to handle the different types of Tags within the game.
- * They can be retrieved directly with the ITagManager BEP, and are also used indirectly when creating a tag with the Tag BEP.
- *
- * @param <T> The element types of this tag.
- */
 @ZenRegister
-@Document("vanilla/api/tag/ITagManager")
-@ZenCodeType.Name("crafttweaker.api.tag.ITagManager")
-public interface ITagManager<T> extends CommandStringDisplayable {
-    
-    @Nonnull
-    Class<T> getElementClass();
-    
-    /**
-     * Retrieves a tag by its name.
-     * Will also be called by the BEP.
-     * <p>
-     * Note that this method does _not_ yet create the tag if it does not exist.
-     * Adding something to the object created by this tag will create it for the game.
-     *
-     * @param name The Resource location of the tag
-     *
-     * @return A Tag object.
-     */
+@Document("vanilla/api/tag/manager/ITagManager")
+@ZenCodeType.Name("crafttweaker.api.tag.manager.ITagManager")
+public interface ITagManager<T> extends CommandStringDisplayable, Iterable<MCTag<T>> {
+
+    @ZenCodeType.Method
+    @ZenCodeType.Getter("tagFolder")
+    default String tagFolder() {
+
+        String tagDir = TagManager.getTagDir(resourceKey());
+
+        // Really not ideal, but I don't see a better way, lets just hope that other mods don't be dumb and add their tags to other folders.
+        if (tagDir.startsWith("tags/")) {
+            tagDir = tagDir.substring("tags/".length());
+        }
+        return tagDir;
+    }
+
     @ZenCodeType.Method
-    default MCTag<T> getTag(String name) {
-        
-        return getTag(new ResourceLocation(name));
+    default MCTag<T> tag(String id) {
+
+        ResourceLocation rl = ResourceLocation.tryParse(id);
+        Objects.requireNonNull(rl, "Invalid Resource Location passed! Given: '" + id + "'");
```

Comment:
If the resource location is invalid, then we will be throwing an exception anyway. Why not create a RL directly and let `ResourceLocationException` bubble up if needed, so that it also gives the reason for the failure?